operator/internal/lifecycle/client.go

```diff
@@ -0,0 +1,387 @@
+	for _, resource := range resources {
+		toDelete[gvkObject{
+			gvk: resource.GetObjectKind().GroupVersionKind(),
+			nn:  client.ObjectKeyFromObject(resource),
+		}] = resource
+	}
+
+	toSync, err := r.simpleResourceRenderer.Render(ctx, owner)
+	if err != nil {
+		return err
+	}
+
+	// attempt to create as many resources in one pass as we can
+	errs := []error{}
+
+	for _, resource := range toSync {
+		if err := r.patchOwnedResource(ctx, owner, resource); err != nil {
+			errs = append(errs, err)
+		}
+		delete(toDelete, gvkObject{
+			gvk: resource.GetObjectKind().GroupVersionKind(),
+			nn:  client.ObjectKeyFromObject(resource),
+		})
+	}
+
+	for _, resource := range toDelete {
+		if err := r.client.Delete(ctx, resource); err != nil {
+			if !k8sapierrors.IsNotFound(err) {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+// FetchExistingAndDesiredPools fetches the existing and desired node pools for a given cluster, returning
+// a tracker that can be used for determining necessary operations on the pools.
+func (r *ResourceClient[T, U]) FetchExistingAndDesiredPools(ctx context.Context, cluster U) (*PoolTracker, error) {
+	pools := NewPoolTracker(cluster.GetGeneration())
+
+	existingPools, err := r.fetchExistingPools(ctx, cluster)
+	if err != nil {
+		return nil, fmt.Errorf("fetching existing pools: %w", err)
+	}
+
+	desired, err := r.nodePoolRenderer.Render(ctx, cluster)
+	if err != nil {
+		return nil, fmt.Errorf("constructing desired pools: %w", err)
+	}
+
+	pools.addExisting(existingPools...)
+	pools.addDesired(desired...)
+
+	return pools, nil
+}
+
+// Builder is an interface for our used methods of *sigs.k8s.io/controller-runtime/pkg/builder.Builder
+type Builder interface {
+	For(object client.Object, opts ...builder.ForOption) *builder.Builder
+	Owns(object client.Object, opts ...builder.OwnsOption) *builder.Builder
+	Watches(object client.Object, eventHandler handler.EventHandler, opts ...builder.WatchesOption) *builder.Builder
+}
+
+// WatchResources configures resource watching for the given cluster, including StatefulSets and other resources.
+func (r *ResourceClient[T, U]) WatchResources(builder Builder, cluster U) error {
+	// set that this is for the cluster
+	builder.For(cluster)
+
+	// set an Owns on node pool statefulsets
+	builder.Owns(&appsv1.StatefulSet{})
+
+	for _, resourceType := range r.simpleResourceRenderer.WatchedResourceTypes() {
+		mapping, err := getResourceScope(r.mapper, r.scheme, resourceType)
+		if err != nil {
+			return err
+		}
+
+		if mapping.Name() == meta.RESTScopeNamespace.Name() {
+			// we're working with a namespace scoped resource, so we can work with ownership
+			builder.Owns(resourceType)
+			continue
+		}
+
+		// since resources are cluster-scoped we need to call a Watch on them with some
+		// custom mappings
+		builder.Watches(resourceType, handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, o client.Object) []reconcile.Request {
```

Comment:
Tricky, I would have reached for filter predicates at first but I see why that doesn't work.